src/main/java/fi/okm/jod/yksilo/controller/profiili/KoulutusController.java

```diff
@@ -24,97 +25,78 @@
     return service.findAll(user);
   }
 
-  @PostMapping(path = "/koulutukset")
+  @PostMapping
   @ResponseStatus(HttpStatus.OK)
   @Operation(
-      summary = "Adds a new set of koulutus and its associated kategoria",
-      description =
-          """
-              This endpoint can be used to add a new Kategoria and its associated Koulutus.
-              If kategoria is not present koulutus will be added without kategoria.
-              """)
-  KoulutusUpdateResultDto create(
-      @RequestBody @Valid KoulutusKategoriaDto dto, @AuthenticationPrincipal JodUser user) {
-    return service.create(user, dto.kategoria(), dto.koulutukset());
+      summary =
+          "Creates a set of Koulutus associated with an optional Kategoria, creating the Kategoria if necessary")
+  KoulutusUpdateResultDto add(
+      @RequestBody @Validated(Add.class) KoulutusKategoriaDto dto,
+      @AuthenticationPrincipal JodUser user) {
+    return service.upsert(user, dto.kategoria(), dto.koulutukset());
   }
 
-  @PatchMapping(path = "/koulutukset/{id}")
+  @PutMapping
   @ResponseStatus(HttpStatus.OK)
   @Operation(
-      summary = "Updates the koulutus by id",
-      description =
-          """
-              This endpoint can be used to update Koulutus.
-              """)
-  KoulutusUpdateResultDto updateKoulutus(
-      @PathVariable UUID id,
-      @RequestBody @Valid KoulutusDto dto,
-      @AuthenticationPrincipal JodUser user) {
-    if (dto.id() == null || !id.equals(dto.id())) {
-      throw new IllegalArgumentException("Invalid identifier");
-    }
-    return service.update(user, dto);
+      summary = "Updates a set of Koulutus",
+      description = "If an existing Koulutus in a Kategoria is omitted, it will be removed.")
+  KoulutusUpdateResultDto update(
+      @RequestBody @Valid KoulutusKategoriaDto dto, @AuthenticationPrincipal JodUser user) {
+    return service.merge(user, dto.kategoria(), dto.koulutukset());
+  }
+
+  @DeleteMapping
+  @ResponseStatus(HttpStatus.NO_CONTENT)
+  void deleteAll(@RequestParam Set<UUID> ids, @AuthenticationPrincipal JodUser user) {
+    service.delete(user, ids);
   }
 
-  @PatchMapping(path = "/kategoriat/{id}")
+  @PutMapping(path = "/{id}")
   @ResponseStatus(HttpStatus.OK)
-  @Operation(
-      summary = "Updates a Kategoria by id",
-      description =
-          """
-              This endpoint can be used to update the kagoria by id.
-              """)
-  KoulutusUpdateResultDto updateKategoria(
+  void updateKoulutus(
       @PathVariable UUID id,
-      @RequestBody @Valid KategoriaDto dto,
+      @RequestBody @Valid KoulutusDto dto,
       @AuthenticationPrincipal JodUser user) {
     if (dto.id() == null || !id.equals(dto.id())) {
       throw new IllegalArgumentException("Invalid identifier");
     }
-    return service.update(user, dto);
+    service.update(user, dto);
   }
 
-  @GetMapping("/koulutukset/{id}")
+  @GetMapping("/{id}")
```

Comment:
It feels a bit wrong to have `GET /api/profiili/koulutukset` return different kind of objects than `GET /api/profiili/koulutukset/{id}`. I think it would be more clear if individual "koulutukset" would be accessed through a different resource name/path than the combo-objects that combine koulutus+kategoria. Perhaps the same way kategoriat now has `PUT /api/profiili/koulutukset/kategoriat/{id}`, we could have `GET/PUT/DELETE /api/profiili/koulutukset/name-for-individual-koulutus-without-kategoria/{id}`.

Problem is name-for-individual-koulutus-without-kategoria is actually also "koulutus" and `GET /api/profiili/koulutukset/koulutukset/{id}` is not very good solution either. This forces us in direction of `GET /api/profiili/koulutus-ja-kategoria-dtot/koulutukset/{id}` which is ugly in it's own way. If only there was a good clear name for koulutus-ja-kategoria-dtot...

Coming up with good names is hard, but anyway, it would IMO be at least a minor improvement in clarity if collection-endpoint resource would be same as single-item-endpoint resource.

Reply:
I don't think that returning different representation for collections is generally "wrong" (quite often some kind of simpler version is used to avoid returning too much data), but here of course combo-object has quite different shape than plain Koulutus (one reason is avoiding repeating Kategoria when returning everything, and one reason is symmetry with collection POST / PUT). The singular Koulutus apis could also return or accept combo-objects with the restriction that only one Koulutus can be present. Not sure if that is "better", though.